precompile/assetsbridge/bridge_out.go

```diff
@@ -0,0 +1,615 @@
+		return nil, fmt.Errorf("failed to send AssetsUnlocked to bridge: %w", err)
+	}
+
+	// finally update the journal entries to propagate the changes
+	// done to the gas token (BTC in our case)
+	balanceDelta, overflow := uint256.FromBig(inputs.Amount)
+	if overflow {
+		return nil, fmt.Errorf("conversion from big.Int to uint256.Int overflowed: %v", inputs.Amount)
+	}
+
+	// only one side of the transfer to update here as we
+	// burnt funds
+	journal := context.Journal()
+	journal.SubBalance(senderAddr, balanceDelta, tracing.BalanceChangeTransfer)
+
+	return assetUnlocked, nil
+}
+
+func (m *BridgeOutMethod) validateAuthorizationLimits(
+	sendAuth *banktypes.SendAuthorization,
+	requestedCoins sdk.Coins,
+) error {
+	if sendAuth.SpendLimit == nil || sendAuth.SpendLimit.Empty() {
+		return fmt.Errorf("no allowance for for %v", requestedCoins[0].Denom)
+	}
+
+	for _, requestedCoin := range requestedCoins {
+		allowedAmount := sendAuth.SpendLimit.AmountOf(requestedCoin.Denom)
+		if allowedAmount.IsZero() {
+			return fmt.Errorf("no allowance for %s", requestedCoin.Denom)
+		}
+		if requestedCoin.Amount.GT(allowedAmount) {
+			return fmt.Errorf(
+				"requested amount %s exceeds allowed amount %s for %s",
+				requestedCoin.Amount,
+				allowedAmount,
+				requestedCoin.Denom,
+			)
+		}
+	}
+
+	// Check allowed list if it exists
+	// It shouldn't be set seeing that we don't really use cosmos-sdk
+	// but just in case?
+	if len(sendAuth.AllowList) > 0 {
+		found := false
+		senderAddrStr := sdk.AccAddress(common.HexToAddress(
+			evmtypes.AssetsBridgePrecompileAddress,
+		).Bytes()).String()
+
+		for _, allowedAddr := range sendAuth.AllowList {
+			if allowedAddr == senderAddrStr {
+				found = true
+				break
+			}
+		}
+		if !found {
+			return fmt.Errorf("recipient address not in authorization allow list")
+		}
+	}
+
+	return nil
+}
+
+// validate applies more extensive validation to the inputs, not only
+// the semantic validation but also verify balances, and tokens etc.
+func (m *BridgeOutMethod) validate(
+	context *precompile.RunContext,
+	inputs *bridgeOutInputs,
+) error {
+	sdkCtx := context.SdkCtx()
+	// first check that the token is either BTC or a supported ERC20
+	if !m.isValidToken(sdkCtx, inputs.Token) {
+		return fmt.Errorf("unsupported token: %v", inputs.Token)
+	}
+
+	spenderAddr := context.MsgSender()
+	if err := m.isAmountSpendableByBridge(sdkCtx, spenderAddr, inputs.Token, inputs.Amount); err != nil {
+		return err
+	}
+
+	return m.isAmountAvailable(sdkCtx, spenderAddr, inputs.Token, inputs.Amount)
+}
+
+func (m *BridgeOutMethod) isAmountSpendableByBridge(
+	sdkCtx sdk.Context,
+	spenderAddr common.Address,
+	token common.Address,
+	amount *big.Int,
+) error {
+	// We use the ERC20 allowance call for both the BTC token
+	// and ERC20s as both implement the ERC20 interface.
+
+	bridgeAddrBytes := common.HexToAddress(
+		evmtypes.AssetsBridgePrecompileAddress,
+	).Bytes()
+
+	call, err := evmtypes.NewERC20AllowanceCall(
```

Comment:
Not sure why we need `from` and `to` here. `allowance` takes only two args: `owner` and `spender`.

Reply:
You're right. I missed the fact this is required to implement the `ContractCall` interface. My bad. Nevermind.